### PR TITLE
add restrictions to migration

### PR DESCRIPTION
**Pull request by jmtroth0**
adding the discussed restrictions on the migration, i.e. old exports and custom reports. 

@snopoke @gcapalbo

### CODE REVIEW

File: corehq/apps/couch_sql_migration/tests/test_migration.py [hunk 1]
```diff
@@ -77,6 +82,21 @@ def _get_case_ids(self):
 
 
 class MigrationTestCase(BaseMigrationTestCase):
+    def test_migration_blacklist(self):
+        COUCH_SQL_MIGRATION_BLACKLIST.set(self.domain_name, True, NAMESPACE_DOMAIN)
+        with self.assertRaises(AssertionError):
+            self._do_migration(self.domain_name)
+        COUCH_SQL_MIGRATION_BLACKLIST.set(self.domain_name, False, NAMESPACE_DOMAIN)
+
+    @patch.object(HQExportSchema, "get_db", MagicMock())
```
Comment (gcapalbo) [inline corehq/apps/couch_sql_migration/tests/test_migration.py]: would it be better to patch `stale_get_export_count` and give it a nonzero return value? i'm not sure i understand how patching get_db() would create the test condition

File: corehq/apps/couch_sql_migration/couchsqlmigration.py [hunk 2]
```diff
@@ -286,6 +285,13 @@ def _diff_ledgers(self, case_ids):
                 filter_ledger_diffs(diffs)
             )
 
+    def _assert_no_migration_restrictions(self, domain_name):
+        assert should_use_sql_backend(domain_name)
+        assert not COUCH_SQL_MIGRATION_BLACKLIST.enabled(domain_name, NAMESPACE_DOMAIN)
+        assert not stale_get_export_count(domain_name)
+        assert not any(custom_report_domain == domain_name
+                       for custom_report_domain in settings.DOMAIN_MODULE_MAP.keys())
```
Comment (gcapalbo) [inline corehq/apps/couch_sql_migration/couchsqlmigration.py]: could there be projects that use custom code that is not custom reports that would also need to be excluded?

**Review by gcapalbo — APPROVED** [review]
a few comments, i don't think anything is blocking

File: corehq/apps/couch_sql_migration/couchsqlmigration.py (hunk 2 above)
Comment (snopoke) [reply corehq/apps/couch_sql_migration/couchsqlmigration.py]: Yes, it's possible that there are others e.g. https://github.com/dimagi/commcare-hq/blob/014d03024d1aa792a84ed911bc385ca1f4ff60fa/custom/hki (though I think that one applies to the bihar domains which are already covered).

File: corehq/apps/couch_sql_migration/couchsqlmigration.py [hunk 3]
```diff
@@ -286,6 +285,13 @@ def _diff_ledgers(self, case_ids):
                 filter_ledger_diffs(diffs)
             )
 
+    def _assert_no_migration_restrictions(self, domain_name):
+        assert should_use_sql_backend(domain_name)
```
[the remark this answers is not in the document]
Comment (jmtroth0) [reply corehq/apps/couch_sql_migration/couchsqlmigration.py]: Prior to this, the migration is setting an override on the local thread to use the sql backend, which makes this true https://github.com/dimagi/commcare-hq/blob/master/corehq/form_processor/utils/general.py#L19

The domain itself should not yet be on the sql backend, but that case is caught earlier

File: corehq/apps/couch_sql_migration/couchsqlmigration.py (hunk 2 above)
Comment (jmtroth0) [reply corehq/apps/couch_sql_migration/couchsqlmigration.py]: @gcapalbo I'd believe that it's possible. What are the chances there is a nice, compiled list of those projects somewhere?

Would it be reasonable/sufficient to take the names from the list of dirs in `custom/`, `custom/_legacy`, `custom/apps`, `custom/reports`, and restrict those as well or do you have something else in mind?

**Review by gcapalbo — APPROVED** [review]

File: corehq/apps/couch_sql_migration/couchsqlmigration.py (hunk 2 above)
Comment (gcapalbo) [reply corehq/apps/couch_sql_migration/couchsqlmigration.py]: i don't think there's a list of projects anywhere, but i think looking at the directories you mentioned and seeing which ones don't already exist in the `DOMAIN_MODULE_MAP` would probably be the additional candidates to look out for. you may need to share the list with the dev team to find out which projects use them

**Pull request merged**